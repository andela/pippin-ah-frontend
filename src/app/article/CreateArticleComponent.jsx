import React from 'react';
import Editor from './EditorComponent';
import './create-article.scss';

const CreateArticleComponent = () => {
  return (
    <div className="create-wrapper">
      <div className="create-container">
        <div className="content-container">
          <div className="create-author-details">
            <div className="author-photo" />
            <div className="create-author-name">
              <div className="create-auth-name">Tom Henkins</div>
            </div>
          </div>
          <div className="create-form">
            <form className="article-form">
              <input type="text" placeholder="Title" />
              <select
                name="category"
                defaultValue=""
                className="browser-default"
                required
              >
                <option value="" disabled>
                  Select Category
                </option>
                <option value="mathematics">Mathematics</option>
                <option value="arts">Arts</option>
                <option value="science">Science</option>
                <option value="engineering">Engineering</option>
                <option value="technology">Technology</option>
              </select>
              <input type="text" placeholder="Description" />
              <div className="create-article-text">
                <Editor />
              </div>
              <div className="create-article-button">
                <button type="submit">POST</button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
};

export default CreateArticleComponent;
