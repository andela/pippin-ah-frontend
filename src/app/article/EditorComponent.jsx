import React from 'react';
import { Editor } from '@tinymce/tinymce-react';
import { uploadImage } from '../util/uploadToCloudinary';

const EditorComponent = ({ handleEditorChange }) => {
  return (
    <Editor
      init={{
        plugins: 'autolink link image lists print preview',
        toolbar:
          'undo redo | bold italic | alignleft aligncenter alignright| link image |',
        width: '100%',
        height: '40%',
        entity_encoding: 'raw',
        automatic_uploads: true,
        file_picker_types: 'file image media',
        file_picker_callback(cb) {
          const input = document.createElement('input');
          input.setAttribute('type', 'file');
          input.setAttribute('accept', 'image/*');
          input.onchange = function onInputChange() {
            const file = this.files[0];
            const reader = new FileReader();
            reader.onload = function onReaderLoad() {
              const id = `blobid${new Date().getTime()}`;
              const { blobCache } = tinymce.activeEditor.editorUpload;
              const base64 = reader.result.split(',')[1];
              const blobInfo = blobCache.create(id, file, base64);
              blobCache.add(blobInfo);
              cb(blobInfo.blobUri(), { title: file.name });
            };
            reader.readAsDataURL(file);
          };
          input.click();
        },
        images_upload_handler(blobInfo, success, failure) {
          const formData = new FormData();
          formData.append('file', blobInfo.blob(), blobInfo.filename());
          uploadImage('article', 'data', formData)
            .then(imageLink => {
              success(imageLink);
            })
            .catch(e => {
              failure(`error uploading imgage: ${e}`);
            });
        },
      }}
      onEditorChange={content => {
        handleEditorChange(content);
      }}
    />
  );
};

export default EditorComponent;
