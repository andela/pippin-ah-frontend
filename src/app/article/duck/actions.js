import types from './types';

const setCreateStatus = createStatus => ({
  type: types.SET_CREATE_STATUS,
  createStatus,
});

const setFetchArticleState = fetchArticleState => ({
  type: types.SET_FETCH_ARTICLE_STATE,
  fetchArticleState,
});

const setFetchArticleError = errorMessage => ({
  type: types.SET_FETCH_ARTICLE_ERROR,
  errorMessage,
});

// const setArticleCategory = articleCategory => ({
//   type: types.SET_ARTICLE_CATEGORY,
//   articleCategory,
// });

const addArticleData = (articleData, articleCategory) => ({
  type: types.ADD_ARTICLE_DATA,
  articleData,
  articleCategory,
});

export default {
  setFetchArticleState,
  setFetchArticleError,
  // setArticleCategory,
  addArticleData,
  setCreateStatus,
};
