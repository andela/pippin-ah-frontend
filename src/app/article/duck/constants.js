export default {
  CREATING: 'CREATING',
  CREATE_ERROR: 'CREATE_ERROR',
  CREATE_SUCCESS: 'CREATE_SUCCESS',
  FETCH_SINGLE_SUCCESS: 'FETCH_SINGLE_SUCCESS',
  FETCH_SINGLE_ERROR: 'FETCH_SINGLE_ERROR',
  FETCHING_SINGLE: 'FETCHING_SINGLE',
  FETCHING_ARTICLE: 'FETCHING_ARTICLE',
  FETCH_ARTICLE_ERROR: 'FETCH_ARTICLE_ERROR',
  FETCH_ARTICLE_SUCCESS: 'FETCH_ARTICLE_SUCCESS',
  HIGHLIGHT_UPLOAD_SUCCESS: 'HIGHLIGHT_UPLOAD_SUCCESS',
  HIGHLIGHT_UPLOADING: 'HIGHLIGHT_UPLOADING',
  HIGHLIGHT_UPLOAD_FAILED: 'HIGHLIGHT_UPLOAD_FAILED',
};
