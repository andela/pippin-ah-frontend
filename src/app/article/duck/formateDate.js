export default rawDate => {
  const date = new Date(rawDate).toDateString();
  return {
    weekDay: date.slice(0, 3),
    month: date.slice(4, 7),
    day: date.slice(8, 10),
    year: date.slice(11, 15),
  };
};
