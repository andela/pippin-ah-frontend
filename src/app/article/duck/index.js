export {
  doCreateArticle,
  doFetchArticle,
  doSetCategory,
  doUpdateCategoryData,
  doFetchArticles,
  doBookmarkArticle,
  doRemoveBookmark,
} from './operations';
export { createArticleReducer, fetchArticleReducer } from './reducers';
export { default as constants } from './constants';
export { default as types } from './types';
export { default as actions } from './actions';
export { default as formatDate } from './formateDate';
