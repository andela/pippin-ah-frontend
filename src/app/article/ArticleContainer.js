import { connect } from 'react-redux';
import ArticleComponent from './ArticleComponent';
import { doFetchArticle, doBookmarkArticle } from './duck';

const mapStateToProps = ({ createArticle: { singleFetchStatus } }) => {
  return { singleFetchStatus };
};
const mapDispatchToProps = dispatch => {
  return {
    fetchSingleArticle: slug => dispatch(doFetchArticle(slug)),
    bookmarkArticle: (slug, articleData) =>
      dispatch(doBookmarkArticle(slug, articleData)),
  };
};

const ArticleContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ArticleComponent);

export default ArticleContainer;
