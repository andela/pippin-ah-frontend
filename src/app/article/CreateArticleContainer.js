import { connect } from 'react-redux';
import { doCreateArticle } from './duck';
import CreateArticleComponent from './CreateArticleComponent';

const mapStateToProps = ({ createArticle: { createStatus } }) => {
  return { createStatus };
};
const mapDispatchToProps = dispatch => {
  return {
    signupUser: articleDetails => dispatch(doCreateArticle(articleDetails)),
  };
};
const ArticleContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(CreateArticleComponent);

export default ArticleContainer;
