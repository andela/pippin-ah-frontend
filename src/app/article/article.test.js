import React from 'react';
import { mount, shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { Redirect } from 'react-router-dom';
import axios from 'axios';
import ArticleContainer from './ArticleContainer';
import CreateArticleContainer from './CreateArticleContainer';
import CreateArticleComponent from './CreateArticleComponent';
import EditorComponent from './EditorComponent';
import PreloaderComponent from '../loaders/PreloaderComponent';
import { actions, types, constants, createArticleReducer } from './duck';

jest.mock('axios');
global.localStorage = {
  getItem: key => {
    return this.store[key] || null;
  },
  setItem: (key, value) => {
    this.store[key] = value.toString();
  },
  removeItem(key) {
    delete this.store[key];
  },
};

describe('ARTICLE TEST SUITE', () => {
  describe('Article Container', () => {
    it('should render the Article Page', () => {
      const component = mount(<ArticleContainer />);
      expect(component.exists()).toBe(true);
      expect(component).toMatchSnapshot();
    });
  });

  describe('CREATE ARTICLE TEST SUITE', () => {
    describe('Create Article Render Components', () => {
      it('should render the CreateArticleContainer', () => {
        const component = shallow(<CreateArticleContainer />);
        expect(component.exists()).toBe(true);
        expect(component).toMatchSnapshot();
      });

      it('should render the CreateArticleComponent', () => {
        const component = shallow(<CreateArticleComponent createStatus="" />);
        expect(component.exists()).toBe(true);
        expect(component).toMatchSnapshot();
      });

      it('should render the EditorComponent', () => {
        const component = shallow(<EditorComponent />);
        expect(component.exists()).toBe(true);
        expect(component).toMatchSnapshot();
      });
    });

    describe('Create Article Simulate Changes', () => {
      it('should redirect to login if user is not authenticated', () => {
        localStorage.removeItem('token');
        const component = shallow(
          <CreateArticleComponent
            createStatus={{
              status: '',
            }}
          />,
        );
        expect(
          component.containsMatchingElement(<Redirect to="/login" />),
        ).toEqual(true);
      });

      it('should call editor change handler when editor is modified', () => {
        localStorage.setItem('token', 'token');
        const handleEditorChange = jest.fn();
        const component = mount(
          <CreateArticleComponent
            createStatus=""
            handleEditorChange={handleEditorChange}
          >
            <EditorComponent handleEditorChange={jest.fn()} />
          </CreateArticleComponent>,
        );
        const spy = jest.spyOn(component.instance(), 'handleEditorChange');
        component
          .find('Editor')
          .props()
          .onEditorChange('new value');
        expect(spy).toHaveBeenCalledWith('new value');
      });

      it('should call createArticle function when form is submitted', () => {
        const createArticle = jest.fn();
        const component = shallow(
          <CreateArticleComponent
            createStatus=""
            createArticle={createArticle}
          />,
        );
        component.find('form').simulate('submit', {
          preventDefault: () => {},
          target: {
            elements: {
              title: { value: 'title' },
              description: { value: 'description' },
              category: { value: 'category' },
            },
          },
        });
        expect(createArticle).toHaveBeenCalledWith({
          title: 'title',
          description: 'description',
          category: 'category',
          body: '',
        });
      });

      it('should scroll to top to display error message when create fails', () => {
        global.scrollTo = jest.fn();
        const component = shallow(
          <CreateArticleComponent
            createStatus={{
              status: constants.CREATE_ERROR,
            }}
          />,
        );
        component.setState({
          canScrollToTop: true,
        });
        expect(global.scrollTo).toHaveBeenCalledWith(0, 0);
      });

      it('should redirect if article creation is successful', () => {
        const component = shallow(
          <CreateArticleComponent
            createStatus={{
              status: constants.CREATE_SUCCESS,
              data: 'new-article',
            }}
          />,
        );
        expect(
          component.containsMatchingElement(
            <Redirect to="/articles/new-article" />,
          ),
        ).toEqual(true);
      });

      it('should render the PreloaderComponent when creating article', () => {
        const component = shallow(
          <CreateArticleComponent
            createStatus={{
              status: constants.CREATING,
            }}
          />,
        );
        expect(
          component.containsMatchingElement(<PreloaderComponent />),
        ).toEqual(true);
      });
    });
  });

  describe('Create Article Actions', () => {
    it('it should set the create status', () => {
      const { setCreateStatus } = actions;
      const createStatus = { status: 'status', data: 'data' };
      const action = setCreateStatus(createStatus);
      expect(action).toEqual({
        type: types.SET_CREATE_STATUS,
        createStatus,
      });
    });
  });

  describe('Create Article Reducers', () => {
    it('should setup default state values', () => {
      const state = createArticleReducer(undefined, {
        type: '@@INIT',
      });
      expect(state.createStatus).toEqual({
        status: '',
        data: '',
      });
    });

    it('should change the create state', () => {
      const action = {
        type: types.SET_CREATE_STATUS,
        createStatus: constants.CREATING,
      };
      const state = createArticleReducer(undefined, action);
      expect(state.createStatus).toEqual(constants.CREATING);
    });
  });

  describe('Connected Create Article Component Dispatches Create Success', () => {
    const initialState = {
      createArticle: {
        createStatus: {
          status: '',
          data: '',
        },
      },
    };
    const mockStore = configureStore([thunk]);
    const store = mockStore(initialState);
    let wrapper;
    beforeEach(() => {
      const response = {
        data: {
          slug: 'new article 001',
        },
      };
      axios.post.mockResolvedValue(response);
      wrapper = mount(
        <Provider store={store}>
          <CreateArticleContainer />
        </Provider>,
      );
      wrapper.find('form').simulate('submit', {
        preventDefault: () => {},
        target: {
          elements: {
            title: { value: 'title' },
            description: { value: 'description' },
            category: { value: 'category' },
          },
        },
      });
    });

    it('should render the connected Create Article component', () => {
      expect(wrapper.find(CreateArticleContainer).length).toEqual(1);
    });

    it('should dispatch create action for successfull article creation', () => {
      const storeActions = store.getActions();
      expect(storeActions[0].type).toEqual('SET_CREATE_STATUS');
    });
  });

  describe('Connected Create Article Component Dispatches Create Error', () => {
    const initialState = {
      createArticle: {
        createStatus: {
          status: '',
          data: '',
        },
      },
    };
    const mockStore = configureStore([thunk]);
    const store = mockStore(initialState);
    let wrapper;
    beforeEach(() => {
      const response = {
        response: {
          data: {
            slug: 'new article 001',
          },
        },
      };
      axios.post.mockImplementation(() => Promise.reject(response));
      wrapper = mount(
        <Provider store={store}>
          <CreateArticleContainer />
        </Provider>,
      );
      wrapper.find('form').simulate('submit', {
        preventDefault: () => {},
        target: {
          elements: {
            title: { value: 'title' },
            description: { value: 'description' },
            category: { value: 'category' },
          },
        },
      });
    });

    it('should dispatch create error for failed article creation', () => {
      const storeActions = store.getActions();
      expect(storeActions[1].createStatus.status).toEqual(
        constants.CREATE_ERROR,
      );
    });
  });
});
