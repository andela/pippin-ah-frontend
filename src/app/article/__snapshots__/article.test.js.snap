// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article Container should render the Article Page 1`] = `
<ArticleContainer>
  <ArticleComponent>
    <div
      className="cover"
    >
      <div
        className="left-sidebar-cover"
      >
        <div
          className="left-sidebar"
        >
          <div
            className="side-bookmark"
          />
          <div
            className="side-like"
          />
          <div
            className="side-facebook"
          />
          <div
            className="side-twitter"
          />
        </div>
      </div>
      <div
        className="container"
      >
        <div
          className="wrapper"
        >
          <div
            className="article-header"
          >
            <div
              className="article-title"
            >
              <p>
                Nature In Her Genes
              </p>
            </div>
            <div
              className="article-description"
            >
              <p>
                A fingerprint of gene activity could reveal the true youthfulness of our kidneys, hearts and muscle, regardless of our biological age. The technique might one day be used.
              </p>
            </div>
          </div>
          <div
            className="article-category"
          >
            <div
              className="category-image"
            />
          </div>
          <div
            className="author-description"
          >
            <div
              className="profile-photo-container"
            />
            <div
              className="description-text"
            >
              <div
                className="author-info"
              >
                <span
                  className="author-name"
                >
                  Sarah Goblin 
                </span>
                <span
                  className="follow-button"
                >
                  follow
                </span>
              </div>
              <div
                className="article-details"
              >
                <span
                  className="article-date"
                >
                  Jan 7, 2019.
                </span>
                <span
                  className="article-read-time"
                >
                  3 min read
                </span>
              </div>
              <div
                className="article-likes"
              >
                <span
                  className="total-likes"
                >
                  21
                </span>
                <div
                  className="like-button"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          className="article-content"
        >
          <div
            className="short-title"
          >
            <h3>
              The Ways of the Man in the Cave
            </h3>
          </div>
          <div
            className="article-text"
          >
            Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words, consectetur, from a Lorem Ipsum passage, and going through the cites of the word in classical literature, discovered the undoubtable source. Lorem Ipsum comes from sections 1.10.32 and 1.10.33 of "de Finibus Bonorum et Malorum" (The Extremes of Good and Evil) by Cicero, written in 45 BC. This book is a treatise on the theory of ethics, very popular during the Renaissance. The first line of Lorem Ipsum, "Lorem ipsum dolor sit amet..", comes from a line in section 1.10.32. The standard chunk of Lorem Ipsum used since the 1500s is reproduced below for those interested. Sections 1.10.32 and 1.10.33 from "de Finibus Bonorum et Malorum" by Cicero are also reproduced in their exact original form, accompanied by English versions from the 1914 translation by H. Rackham.
          </div>
        </div>
        <div
          className="left-sidebar-down"
        >
          <div
            className="side-bookmark"
          />
          <div
            className="side-like"
          />
          <div
            className="side-facebook"
          />
          <div
            className="side-twitter"
          />
        </div>
        <div
          className="comment-section"
        >
          <span
            className="total-comments"
          >
            12
          </span>
          Comments
        </div>
        <div
          className="comment-box"
        >
          <div
            className="header-wrapper"
          >
            <div
              className="comment-owner"
            >
              <div
                className="owner-image"
              />
            </div>
            <div
              className="comment-container"
            >
              <div
                className="comment-header"
              >
                <span
                  className="comment-author"
                >
                  Paul Mayer
                </span>
                <span
                  className="comment-date"
                >
                  January 7, 2019
                </span>
                <span
                  className="reply-button"
                >
                  Reply
                </span>
              </div>
              <p
                className="comment-text"
              >
                Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words, consectetur, from a Lorem Ipsum passage, and going through the cites of the word in classical literature, discovered the undoubtable source. Lorem Ipsum comes from sections 1.10.32 and 1.10.33.
              </p>
            </div>
          </div>
        </div>
        <div
          className="add-comment-box"
        >
          <div
            className="comment-header"
          >
            <div
              className="pen-image"
            />
            <span
              className="comment-author"
            >
              Write Comment
            </span>
          </div>
          <textarea
            defaultValue=""
            placeholder="Write your comment..."
          />
          <button
            type="button"
          >
            Post
          </button>
        </div>
        <div
          className="right-sidebar-down"
        >
          <p
            className="more-articles"
          >
            More Articles by
          </p>
          <div
            className="sidebar-author"
          >
            SARAH GOBLIN
          </div>
          <div
            className="sidebar-articles"
          >
            <p>
              Ethical Hacking Simplified
            </p>
            <p>
              True Words of Wisdom
            </p>
            <p>
              Understanding Politics
            </p>
            <p>
              In the Mystery of Life
            </p>
            <p>
              Communication Made Easy
            </p>
            <p>
              In the Days of Mr X
            </p>
            <p>
              Knowing your True Value
            </p>
          </div>
        </div>
      </div>
      <div
        className="right-sidebar-cover"
      >
        <div
          className="right-sidebar"
        >
          <p
            className="more-articles"
          >
            More Articles by
          </p>
          <div
            className="sidebar-author"
          >
            SARAH GOBLIN
          </div>
          <div
            className="sidebar-articles"
          >
            <p>
              Ethical Hacking Simplified
            </p>
            <p>
              True Words of Wisdom
            </p>
            <p>
              Understanding Politics
            </p>
            <p>
              In the Mystery of Life
            </p>
            <p>
              Communication Made Easy
            </p>
            <p>
              In the Days of Mr X
            </p>
            <p>
              Knowing your True Value
            </p>
          </div>
        </div>
      </div>
    </div>
  </ArticleComponent>
</ArticleContainer>
`;
